src/useStorageValue/useStorageValue.ts

```diff
@@ -1,336 +1,267 @@
+      ),
+    remove: () => storage.removeItem(key),
+    store: (val: Type): string | null => {
       const stringified = stringify(val);
 
-      if (stringified) {
+      if (stringified !== null) {
         storage.setItem(key, stringified);
-        return true;
       }
 
-      return false;
-    },
-    removeVal: () => {
-      storage.removeItem(key);
-    },
-    setVal: (val: string | null) => {
-      setState(parse(val, defaultValue) as T);
-    },
-    fetchState: () => {
-      const newVal = methods.current.fetchVal() as T;
-      setState(newVal);
-
-      return newVal !== stateRef.current ? newVal : null;
-    },
-    setState: (nextState: T | null) => {
-      setState(nextState === null ? defaultValue : nextState);
+      return stringified;
     },
   });
 
   const isFirstMount = useFirstMountState();
-  const [state, setState] = useSafeState<T | null | undefined>(
-    initializeWithStorageValue && isFirstMount ? (methods.current.fetchVal() as T) : undefined
+  const [state, setState] = useState<Type | null | undefined>(
+    options?.initializeWithValue && isFirstMount ? storageActions.current.fetch() : undefined
   );
-  const prevState = usePrevious(state);
   const stateRef = useSyncedRef(state);
-  const keyRef = useSyncedRef(key);
-  const isolatedRef = useSyncedRef(isolated);
-
-  // fetch value on mount for the case `initializeWithStorageValue` is false,
-  // effects are not invoked during SSR, so there is no need to check isBrowser here
-  useMountEffect(() => {
-    if (!initializeWithStorageValue) {
-      methods.current.fetchState();
-    }
-  });
 
-  // store default value if it is not null and options configured to store default value
-  useConditionalEffect(
-    () => {
-      methods.current.storeVal(defaultValue as T);
+  const stateActions = useSyncedRef({
+    fetch: () => setState(storageActions.current.fetch()),
+    setRawVal: (val: string | null) => {
+      setState(parse(val, optionsRef.current.defaultValue));
     },
-    undefined,
-    [prevState !== state, storeDefaultValue && state === defaultValue && defaultValue !== null]
-  );
+  });
 
-  // refetch value when key changed
   useUpdateEffect(() => {
-    methods.current.fetchState();
+    stateActions.current.fetch();
   }, [key]);
 
-  // subscribe hook for storage events
-  useIsomorphicLayoutEffect(() => {
-    if (!handleStorageEvent) return;
-
-    // eslint-disable-next-line unicorn/consistent-function-scoping
-    const storageHandler = (ev: StorageEvent) => {
-      if (ev.storageArea !== storage) return;
-      if (ev.key !== keyRef.current) return;
-
-      methods.current.setVal(ev.newValue);
-    };
-
-    on(window, 'storage', storageHandler, { passive: true });
-
-    return () => {
-      off(window, 'storage', storageHandler);
-    };
+  useEffect(() => {
+    if (!options?.initializeWithValue) {
+      stateActions.current.fetch();
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I noticed that here we don't actually need to disable this lint rule since `stateActions` is a ref (ESLint knows that refs don't have to be included in the dependency array, but it does not know that `stateActions` is a ref because it comes from our own custom hook) and `options?.initializeWithValue` does not really change unless the developer makes it change by controlling it with a prop coming from the parent (which might be a rare but valid usecase that we are blocking currently).

Reply:
It should only be used once on first render, therefore there is no legit scenario this hook should be retriggered on option change.

Reply:
Perhaps it isn't. It certainly would be odd and rare 🤔